### ISSUE TITLE
forbid time.Sleep when a context is present

### ISSUE THREAD

**Issue by ccoVeille** [go-critic/go-critic#1477]
I would recommend a new checker about detecting the usage of context when dealing with time.Sleep or time.After usage

For me, this pieces of code should be reported:

```go
func foo(ctx context.Context) {

    // whatever

   <-time.After(3 * time.Second)

    // whatever
}
```

```go
func foo(ctx context.Context) {

    // whatever

    time.Sleep(3 * time.Second)

    // whatever
}
```

Something like this should be suggested

```go
func foo(ctx context.Context) {

    // whatever

   select {
   case <-ctx.Done:
      // do something
   case <-time.After(3 * time.Second):
   }

    // whatever
}
```

What is to be done in the ctx.Done case is something that purely depends on the code.

either a return (returning an error if it's in the signature), consider we can stop waiting

_Originally posted by @ccoVeille in https://github.com/go-critic/go-critic/issues/1474#issuecomment-2747758907_

**Comment by ccoVeille** [comment]
Interesting feedback I got when I opened an issue in [contextcheck](https://github.com/kkHAIKE/contextcheck) about the same thing. I had forgotten I already discussed about my time.Sleep linter idea in go-critic.

> Based on my experience, it's entirely feasible to avoid using time.Sleep in backend services and instead use time.After; at least that's what I do.
>
> Secondly, in Go's SSA analysis, it's difficult to determine the execution order of instructions. Complex block analysis logic is required to identify context-independent functions like time.Sleep(if time.Sleep is called after the use of context, then no warning is necessary.). On the other hand, `contextcheck` is built on an analysis mechanism that traces `context` reference chains.
>
> In conclusion, I recommend using a linter to completely prohibit the use of time.Sleep.
> 
> _Originally posted by @kkHAIKE in https://github.com/kkHAIKE/contextcheck/issues/29#issuecomment-2889351684_
